<a href="https://www.kaggle.com/code/mrafraim/sales-analytics-data-profiling-and-preparation?scriptVersionId=344581596" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Sales Analytics - Data Profiling and Preparation

# 1. Data Understanding & Profiling 

# 1.1 Load Raw Data

This section loads the original workbook and performs an initial inspection of its structure.

The objective is to understand the available tables and establish the starting point of the analysis before applying any transformations or cleaning.

In [1]:
import pandas as pd

# Show every column without truncation
pd.set_option('display.max_columns', None)

In [2]:
# File path
file_path = "/kaggle/input/datasets/shandeep777/retail-supply-chain-sales-dataset/Retail-Supply-Chain-Sales-Dataset.xlsx"

# Load raw tables
sales_df = pd.read_excel(
    file_path,
    sheet_name="Retails Order Full Dataset"
)

calendar_df = pd.read_excel(
    file_path,
    sheet_name="Calendar Table"
)

In [3]:
# Display dataset dimensions
print(f"Sales dataset: {sales_df.shape[0]:,} rows × {sales_df.shape[1]} columns")
print(f"Calendar table: {calendar_df.shape[0]:,} rows × {calendar_df.shape[1]} columns")

Sales dataset: 9,994 rows × 23 columns
Calendar table: 1,826 rows × 12 columns


In [4]:
# Preview the raw tables
display(sales_df.head())
display(calendar_df.head())

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Retail Sales People,Product ID,Category,Sub-Category,Product Name,Returned,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,2016-08-11,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,Cassandra Brandow,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,Not,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,2016-08-11,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,Cassandra Brandow,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",Not,731.9400,3,0.00,219.5820
2,3,CA-2016-138688,2016-12-06,2016-12-06,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,Anna Andreadi,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,Not,14.6200,2,0.00,6.8714
3,4,US-2015-108966,2015-11-10,2015-11-10,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,Cassandra Brandow,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,Not,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,2015-11-10,2015-11-10,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,Cassandra Brandow,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,Not,22.3680,2,0.20,2.5164


,Date,Year,Quarter,Quarter (Q),Quarter & Year,Month,Month Name,Month & Year,Week of Year,Week of Year (W),Day of Week,Day Name
0,2016-07-01,2016,3,Q3,2016 - Q3,7,Jul,Jul 2016,27,Week 27,5,Friday
1,2016-07-02,2016,3,Q3,2016 - Q3,7,Jul,Jul 2016,27,Week 27,6,Saturday
2,2016-07-03,2016,3,Q3,2016 - Q3,7,Jul,Jul 2016,27,Week 27,7,Sunday
3,2016-07-04,2016,3,Q3,2016 - Q3,7,Jul,Jul 2016,28,Week 28,1,Monday
4,2016-07-05,2016,3,Q3,2016 - Q3,7,Jul,Jul 2016,28,Week 28,2,Tuesday


In [5]:
# List available columns
print("Sales Dataset Columns:")
print(sales_df.columns.tolist())

print("\nCalendar Table Columns:")
print(calendar_df.columns.tolist())

Sales Dataset Columns:
['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Retail Sales People', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Returned', 'Sales', 'Quantity', 'Discount', 'Profit']

Calendar Table Columns:
['Date', 'Year', 'Quarter', 'Quarter (Q)', 'Quarter & Year', 'Month', 'Month Name', 'Month & Year', 'Week of Year', 'Week of Year (W)', 'Day of Week', 'Day Name']


# 1.2 Structural Inspection

We inspect the structure, data types, and completeness of the raw tables before applying any transformations.

This helps identify fields that may require type conversion or further validation during the data preparation stage.

In [6]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Row ID               9994 non-null   int64         
 1   Order ID             9994 non-null   object        
 2   Order Date           9994 non-null   datetime64[ns]
 3   Ship Date            9994 non-null   datetime64[ns]
 4   Ship Mode            9994 non-null   object        
 5   Customer ID          9994 non-null   object        
 6   Customer Name        9994 non-null   object        
 7   Segment              9994 non-null   object        
 8   Country              9994 non-null   object        
 9   City                 9994 non-null   object        
 10  State                9994 non-null   object        
 11  Postal Code          9994 non-null   int64         
 12  Region               9994 non-null   object        
 13  Retail Sales People  9994 non-nul

In [7]:
calendar_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1826 entries, 0 to 1825
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              1826 non-null   datetime64[ns]
 1   Year              1826 non-null   int64         
 2   Quarter           1826 non-null   int64         
 3   Quarter (Q)       1826 non-null   object        
 4   Quarter & Year    1826 non-null   object        
 5   Month             1826 non-null   int64         
 6   Month Name        1826 non-null   object        
 7   Month & Year      1826 non-null   object        
 8   Week of Year      1826 non-null   int64         
 9   Week of Year (W)  1826 non-null   object        
 10  Day of Week       1826 non-null   int64         
 11  Day Name          1826 non-null   object        
dtypes: datetime64[ns](1), int64(5), object(6)
memory usage: 171.3+ KB


## Column-Level Overview

A column-level summary is created to assess data types, completeness, and cardinality for the raw dataset.

In [8]:
sales_profile = pd.DataFrame({
    "Column": sales_df.columns,
    "Data Type": sales_df.dtypes.astype(str).values,
    "Non-Null": sales_df.notna().sum().values,
    "Unique Values": sales_df.nunique().values
})

display(sales_profile)

,Column,Data Type,Non-Null,Unique Values
0,Row ID,int64,9994,9994
1,Order ID,object,9994,5009
2,Order Date,datetime64[ns],9994,1237
3,Ship Date,datetime64[ns],9994,1215
4,Ship Mode,object,9994,4
5,Customer ID,object,9994,793
6,Customer Name,object,9994,793
7,Segment,object,9994,3
8,Country,object,9994,1
9,City,object,9994,531


In [9]:
calendar_profile = pd.DataFrame({
    "Column": calendar_df.columns,
    "Data Type": calendar_df.dtypes.astype(str).values,
    "Non-Null": calendar_df.notna().sum().values,
    "Unique Values": calendar_df.nunique().values
})

display(calendar_profile)

,Column,Data Type,Non-Null,Unique Values
0,Date,datetime64[ns],1826,1826
1,Year,int64,1826,5
2,Quarter,int64,1826,4
3,Quarter (Q),object,1826,4
4,Quarter & Year,object,1826,20
5,Month,int64,1826,12
6,Month Name,object,1826,12
7,Month & Year,object,1826,60
8,Week of Year,int64,1826,53
9,Week of Year (W),object,1826,53


## Initial Structural Findings

- The primary dataset contains 9,994 transaction-line records across 23 columns.
- Order Date and Ship Date are already stored as datetime fields, while Sales, Quantity, Discount, and Profit are stored as numeric fields.
- Row ID is unique across all records and can serve as the transaction-line identifier.
- Order ID is repeated across records, indicating that the dataset is recorded at order-line level rather than order level.
- Customer ID identifies 793 unique customers, while Product ID identifies 1,862 unique products.
- Postal Code is stored numerically but represents a geographic identifier rather than a quantitative measure.
- Country contains only one unique value and therefore has limited analytical value for this project.
- The Calendar Table contains one unique record per date and can potentially serve as the basis for the report's date dimension.
- A difference between the number of unique Product IDs and Product Names was identified and will be investigated during the data-quality assessment.

# 1.3 Identifier & Grain Validation

## Row ID Uniqueness

`Row ID` is expected to uniquely identify each transaction-line record. We validate its uniqueness before using it as the record-level identifier.

In [10]:
row_id_duplicates = sales_df["Row ID"].duplicated().sum()

print(f"Duplicate Row IDs: {row_id_duplicates:,}")

Duplicate Row IDs: 0


In [11]:
print(f"Total records: {len(sales_df):,}")
print(f"Unique Row IDs: {sales_df['Row ID'].nunique():,}")

Total records: 9,994
Unique Row IDs: 9,994


**Finding:** `Row ID` is unique across all 9,994 records, confirming that it can serve as the record-level identifier for the transactional dataset.

## Order-Level vs Transaction-Line Grain

`Order ID` is expected to repeat when a single order contains multiple products or transaction lines. We examine the distribution of transaction lines per order to validate the grain of the dataset.

In [12]:
order_line_counts = sales_df.groupby("Order ID").size()

print(f"Unique Orders: {order_line_counts.size:,}")
print(f"Orders with multiple lines: {(order_line_counts > 1).sum():,}")
print(f"Maximum lines in a single order: {order_line_counts.max():,}")

Unique Orders: 5,009
Orders with multiple lines: 2,471
Maximum lines in a single order: 14


In [13]:
multi_line_orders = (
    order_line_counts[order_line_counts > 1]
    .sort_values(ascending=False)
)

display(multi_line_orders.head(10))

Order ID
CA-2017-100111    14
CA-2017-157987    12
US-2016-108504    11
CA-2016-165330    11
US-2015-126977    10
CA-2016-105732    10
CA-2015-131338    10
CA-2015-158421     9
CA-2015-132626     9
CA-2014-106439     9
dtype: int64

**Finding:** `Order ID` is not unique because individual orders can contain multiple transaction lines. Of the 5,009 orders, 2,471 contain more than one line, with a maximum of 14 lines per order. This confirms that the dataset is recorded at the transaction-line level rather than the order level.

## Customer Identifier Consistency

We validate whether each `Customer ID` consistently maps to a single `Customer Name`. This helps determine whether customer attributes can be reliably separated into a customer dimension during data modeling.

In [14]:
customer_name_counts = (
    sales_df.groupby("Customer ID")["Customer Name"]
    .nunique()
)

inconsistent_customers = customer_name_counts[
    customer_name_counts > 1
]

print(f"Customers with multiple names: {len(inconsistent_customers):,}")

Customers with multiple names: 0


**Finding:** Each `Customer ID` maps consistently to a single `Customer Name`, with no conflicting customer-name mappings identified. This supports separating customer attributes into a dedicated customer dimension during data modeling.

## Product Identifier Consistency

We validate whether each `Product ID` consistently maps to a single `Product Name`. This check is important before creating a product dimension because inconsistent mappings could introduce duplicate or ambiguous product records.

In [15]:
product_name_counts = (
    sales_df.groupby("Product ID")["Product Name"]
    .nunique()
)

inconsistent_products = product_name_counts[
    product_name_counts > 1
]

print(f"Products with multiple names: {len(inconsistent_products):,}")

Products with multiple names: 32


### Investigate Inconsistent Product Mappings

The previous validation identified Product IDs associated with multiple Product Names. We inspect the affected records to determine whether the inconsistencies represent data-quality issues or legitimate product variations.

In [16]:
# Identify Product IDs associated with multiple Product Names
inconsistent_product_ids = inconsistent_products.index

product_mapping_issues = (
    sales_df[
        sales_df["Product ID"].isin(inconsistent_product_ids)
    ][
        ["Product ID", "Product Name", "Category", "Sub-Category"]
    ]
    .drop_duplicates()
    .sort_values(["Product ID", "Product Name"])
)

display(product_mapping_issues.head(20))

,Product ID,Product Name,Category,Sub-Category
2115,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases
2471,FUR-BO-10002213,"Sauder Forest Hills Library, Woodland Oak Finish",Furniture,Bookcases
128,FUR-CH-10001146,"Global Task Chair, Black",Furniture,Chairs
66,FUR-CH-10001146,"Global Value Mid-Back Manager's Chair, Gray",Furniture,Chairs
1459,FUR-FU-10001473,DAX Wood Document Frame,Furniture,Furnishings
2204,FUR-FU-10001473,"Eldon Executive Woodline II Desk Accessories, ...",Furniture,Furnishings
3831,FUR-FU-10004017,"Executive Impressions 13"" Chairman Wall Clock",Furniture,Furnishings
234,FUR-FU-10004017,Tenex Contemporary Contur Chairmats for Low an...,Furniture,Furnishings
1385,FUR-FU-10004091,"Eldon 200 Class Desk Accessories, Black",Furniture,Furnishings
293,FUR-FU-10004091,"Howard Miller 13"" Diameter Goldtone Round Wall...",Furniture,Furnishings


### Investigate the pattern

In [17]:
example_product_id = inconsistent_product_ids[0]

display(
    sales_df[
        sales_df["Product ID"] == example_product_id
    ][
        [
            "Order ID",
            "Order Date",
            "Product ID",
            "Product Name",
            "Category",
            "Sub-Category",
            "Sales",
            "Quantity",
            "Profit"
        ]
    ].sort_values("Order Date")
)

,Order ID,Order Date,Product ID,Product Name,Category,Sub-Category,Sales,Quantity,Profit
9395,CA-2014-125997,2014-09-20,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,801.568,2,-10.0196
6535,CA-2014-128209,2014-11-17,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,4007.840,10,-50.0980
8712,US-2015-145422,2015-03-12,FUR-BO-10002213,"Sauder Forest Hills Library, Woodland Oak Finish",Furniture,Bookcases,359.499,3,-29.6058
2808,CA-2015-148635,2015-07-25,FUR-BO-10002213,"Sauder Forest Hills Library, Woodland Oak Finish",Furniture,Bookcases,704.900,5,56.3920
5918,US-2015-126977,2015-09-17,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,2003.920,5,-25.0490
2115,CA-2015-164882,2015-10-31,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,425.833,1,20.0392
2471,US-2016-135923,2016-01-22,FUR-BO-10002213,"Sauder Forest Hills Library, Woodland Oak Finish",Furniture,Bookcases,451.136,4,-67.6704
9649,CA-2016-107104,2016-11-26,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,3406.664,8,160.3136
9583,CA-2017-116127,2017-06-25,FUR-BO-10002213,DMI Eclipse Executive Suite Bookcases,Furniture,Bookcases,400.784,1,-5.0098
5079,US-2017-133312,2017-11-25,FUR-BO-10002213,"Sauder Forest Hills Library, Woodland Oak Finish",Furniture,Bookcases,359.499,3,-29.6058


**Finding:** 32 Product IDs were identified with multiple Product Name mappings. A detailed inspection of `FUR-BO-10002213` showed that two different product names occur under the same Product ID across overlapping periods, indicating a source-data integrity issue rather than a simple historical product-name change. The raw values will be preserved, and the appropriate treatment will be determined during the data preparation and modeling stages.

### Business Impact of Product Mapping Issues

We quantify the sales and profit associated with Product IDs that have inconsistent Product Name mappings to assess the potential business impact of the source-data issue.

In [18]:
# Total sales and profit
total_sales = sales_df["Sales"].sum()
total_profit = sales_df["Profit"].sum()

# Records associated with inconsistent Product IDs
affected_sales_df = sales_df[
    sales_df["Product ID"].isin(inconsistent_product_ids)
]

affected_sales = affected_sales_df["Sales"].sum()
affected_profit = affected_sales_df["Profit"].sum()

print(f"Total Sales: ${total_sales:,.2f}")
print(f"Affected Sales: ${affected_sales:,.2f}")
print(f"Affected Sales %: {(affected_sales / total_sales) * 100:.2f}%")

print(f"\nTotal Profit: ${total_profit:,.2f}")
print(f"Affected Profit: ${affected_profit:,.2f}")
print(f"Affected Profit %: {(affected_profit / total_profit) * 100:.2f}%")

Total Sales: $2,297,200.86
Affected Sales: $97,465.67
Affected Sales %: 4.24%

Total Profit: $286,397.02
Affected Profit: $13,230.92
Affected Profit %: 4.62%


**Business Impact:** The 32 Product IDs with inconsistent Product Name mappings are associated with $97,465.67 in sales, representing 4.24% of total sales. These records also account for $13,230.92 in profit, or 4.62% of total profit. The issue is therefore material enough to document and consider during data modeling, but it does not invalidate the dataset for overall commercial analysis.

## Customer Attribute Consistency

We validate whether customer-level attributes remain consistent for each Customer ID. This helps identify potential dimensional integrity issues before creating the customer dimension.

In [19]:
customer_attributes = [
    "Customer Name",
    "Segment",
    "Country",
    "City",
    "State",
    "Postal Code",
    "Region"
]

customer_consistency = (
    sales_df.groupby("Customer ID")[customer_attributes]
    .nunique()
)

inconsistent_customer_attributes = (
    customer_consistency > 1
).sum()

display(
    inconsistent_customer_attributes[
        inconsistent_customer_attributes > 0
    ]
    .sort_values(ascending=False)
    .rename("Customers with Multiple Values")
    .to_frame()
)

,Customers with Multiple Values
City,780
Postal Code,780
State,778
Region,766


### Investigating Customer Location Variations

Several customer-level geographic attributes vary across transactions for the same Customer ID. We investigate individual customers to determine whether these fields represent customer master attributes or transaction-level locations.

In [20]:
# Find a customer with multiple cities
customers_with_multiple_cities = (
    customer_consistency[
        customer_consistency["City"] > 1
    ].index
)

example_customer_id = customers_with_multiple_cities[0]

display(
    sales_df[
        sales_df["Customer ID"] == example_customer_id
    ][
        [
            "Order ID",
            "Order Date",
            "Customer ID",
            "Customer Name",
            "City",
            "State",
            "Postal Code",
            "Region"
        ]
    ].sort_values("Order Date")
)

,Order ID,Order Date,Customer ID,Customer Name,City,State,Postal Code,Region
2229,CA-2014-128055,2014-03-31,AA-10315,Alex Avila,San Francisco,California,94122,West
2230,CA-2014-128055,2014-03-31,AA-10315,Alex Avila,San Francisco,California,94122,West
7468,CA-2014-138100,2014-09-15,AA-10315,Alex Avila,New York City,New York,10011,East
7469,CA-2014-138100,2014-09-15,AA-10315,Alex Avila,New York City,New York,10011,East
1299,CA-2015-121391,2015-04-10,AA-10315,Alex Avila,San Francisco,California,94109,West
5198,CA-2016-103982,2016-03-03,AA-10315,Alex Avila,Round Rock,Texas,78664,Central
5199,CA-2016-103982,2016-03-03,AA-10315,Alex Avila,Round Rock,Texas,78664,Central
5200,CA-2016-103982,2016-03-03,AA-10315,Alex Avila,Round Rock,Texas,78664,Central
5201,CA-2016-103982,2016-03-03,AA-10315,Alex Avila,Round Rock,Texas,78664,Central
1159,CA-2017-147039,2017-06-29,AA-10315,Alex Avila,Minneapolis,Minnesota,55407,Central


**Finding:** Customer geographic attributes are not stable across transactions. For example, Customer ID `AA-10315` appears across multiple cities, states, postal codes, and regions during the analysis period. This indicates that these geographic fields represent transaction-level/order-location information rather than fixed customer master attributes. Therefore, they should not be treated as permanent attributes of the customer dimension during data modeling.

We should **not** build our future `DimCustomer` like this:

```text
DimCustomer
├── Customer ID
├── Customer Name
├── City       ❌
├── State      ❌
├── Postal Code ❌
└── Region     ❌
```

Instead, we'll likely separate the concepts:

```text
DimCustomer
├── Customer ID
├── Customer Name
└── Segment

DimGeography
├── City
├── State
├── Postal Code
└── Region
```

with geography associated with the **transaction**.

We aren't building that model yet, but this profiling result is now giving us a reason for the architecture.

## Product Name-to-ID Consistency

The previous validation identified Product IDs associated with multiple Product Names. We now test the reverse relationship to determine whether a Product Name can also be associated with multiple Product IDs.

In [21]:
product_id_counts = (
    sales_df.groupby("Product Name")["Product ID"]
    .nunique()
)

inconsistent_product_names = product_id_counts[
    product_id_counts > 1
]

print(
    f"Product Names associated with multiple Product IDs: "
    f"{len(inconsistent_product_names):,}"
)

Product Names associated with multiple Product IDs: 16


In [22]:
display(
    sales_df[
        sales_df["Product Name"].isin(inconsistent_product_names.index)
    ][
        ["Product ID", "Product Name", "Category", "Sub-Category"]
    ]
    .drop_duplicates()
    .sort_values(["Product Name", "Product ID"])
)

,Product ID,Product Name,Category,Sub-Category
985,OFF-EN-10000461,"#10- 4 1/8"" x 9 1/2"" Recycled Envelopes",Office Supplies,Envelopes
1989,OFF-EN-10000781,"#10- 4 1/8"" x 9 1/2"" Recycled Envelopes",Office Supplies,Envelopes
2531,OFF-BI-10000829,Avery Non-Stick Binders,Office Supplies,Binders
336,OFF-BI-10004140,Avery Non-Stick Binders,Office Supplies,Binders
34,OFF-PA-10000249,Easy-staple paper,Office Supplies,Paper
1080,OFF-PA-10000349,Easy-staple paper,Office Supplies,Paper
283,OFF-PA-10000474,Easy-staple paper,Office Supplies,Paper
4360,OFF-PA-10000565,Easy-staple paper,Office Supplies,Paper
1225,OFF-PA-10001685,Easy-staple paper,Office Supplies,Paper
4999,OFF-PA-10002764,Easy-staple paper,Office Supplies,Paper


**Finding:** Product identity is not consistently represented in the raw dataset. 32 Product IDs are associated with multiple Product Names, while 16 Product Names are associated with multiple Product IDs. Therefore, neither Product ID nor Product Name provides a perfectly consistent one-to-one product mapping in the source data. Product ID will be retained as the primary product identifier for further investigation, while Product Name will not be treated as a unique key.

## Customer Segment

In [23]:
customer_segment_counts = (
    sales_df.groupby("Customer ID")["Segment"]
    .nunique()
)

inconsistent_segments = customer_segment_counts[
    customer_segment_counts > 1
]

print(
    f"Customers with multiple segments: "
    f"{len(inconsistent_segments):,}"
)

Customers with multiple segments: 0


**Finding:** Each `Customer ID` maps to a single `Segment`, with no customers associated with multiple segment classifications. This supports treating `Segment` as a customer-level attribute in the analytical model.

### Our Current Understanding after Section 3
                    
                    RAW TRANSACTION DATA

                        Row ID
                          │
                   Unique transaction
                          │
                       Order ID
                          │
                  Multiple lines/order
                          │
          ┌───────────────┴───────────────┐
          │                               │
      Customer ID                     Product ID
          │                               │
     Customer Name                  Product Name
     Segment                        ⚠️ inconsistent
          │
     Geography
          │
    ⚠️ transaction-level

# 1.4 Completeness Assessment

We assess the completeness of the raw transactional and calendar datasets by identifying missing values at the column level.

Missing-value analysis is performed before transformation to determine whether any fields require imputation, removal, or other treatment.

In [24]:
missing_sales = pd.DataFrame({
    "Column": sales_df.columns,
    "Missing Values": sales_df.isna().sum().values
})

missing_sales["Missing %"] = (
    missing_sales["Missing Values"] / len(sales_df) * 100
)

display(missing_sales)

,Column,Missing Values,Missing %
0,Row ID,0,0.0
1,Order ID,0,0.0
2,Order Date,0,0.0
3,Ship Date,0,0.0
4,Ship Mode,0,0.0
5,Customer ID,0,0.0
6,Customer Name,0,0.0
7,Segment,0,0.0
8,Country,0,0.0
9,City,0,0.0


In [25]:
missing_calendar = pd.DataFrame({
    "Column": calendar_df.columns,
    "Missing Values": calendar_df.isna().sum().values
})

missing_calendar["Missing %"] = (
    missing_calendar["Missing Values"] / len(calendar_df) * 100
)

display(missing_calendar)

,Column,Missing Values,Missing %
0,Date,0,0.0
1,Year,0,0.0
2,Quarter,0,0.0
3,Quarter (Q),0,0.0
4,Quarter & Year,0,0.0
5,Month,0,0.0
6,Month Name,0,0.0
7,Month & Year,0,0.0
8,Week of Year,0,0.0
9,Week of Year (W),0,0.0


**Finding:** No missing values were identified in either the transactional dataset or the calendar table. All fields are complete at the raw-data level, so no missing-value imputation or null-removal step is required.

# 1.5 Duplicate Record Assessment

We assess whether the transactional dataset contains exact duplicate records.

Because the dataset is recorded at transaction-line level, repeated Order IDs are expected and are not treated as duplicates by themselves. The objective here is to identify records that are identical across all available fields.

In [26]:
duplicate_rows = sales_df.duplicated().sum()

print(f"Exact duplicate rows: {duplicate_rows:,}")

Exact duplicate rows: 0


In [27]:
duplicate_percentage = (
    duplicate_rows / len(sales_df) * 100
)

print(f"Duplicate percentage: {duplicate_percentage:.2f}%")

Duplicate percentage: 0.00%


**Finding:** No exact duplicate rows were identified in the transactional dataset. The 9,994 records therefore represent distinct transaction-line records based on the complete set of available fields.

# 1.6 Range & Validity Checks

We evaluate key numerical and date fields for unexpected values or ranges that could affect downstream analysis.

The objective is to identify potential business-rule violations before transformation and modeling.

In [28]:
numeric_summary = sales_df[
    ["Sales", "Profit", "Quantity", "Discount"]
].agg(["min", "max", "mean", "median"])

display(numeric_summary)

,Sales,Profit,Quantity,Discount
min,0.444000,-6599.978000,1.000000,0.000000
max,22638.480000,8399.976000,14.000000,0.800000
mean,229.858001,28.656896,3.789574,0.156203
median,54.490000,8.666500,3.000000,0.200000


In [29]:
validity_checks = {
    "Sales <= 0": (sales_df["Sales"] <= 0).sum(),
    "Quantity <= 0": (sales_df["Quantity"] <= 0).sum(),
    "Discount < 0": (sales_df["Discount"] < 0).sum(),
    "Discount > 1": (sales_df["Discount"] > 1).sum()
}

display(pd.Series(validity_checks, name="Records"))

Sales <= 0       0
Quantity <= 0    0
Discount < 0     0
Discount > 1     0
Name: Records, dtype: int64

In [30]:
date_summary = sales_df[
    ["Order Date", "Ship Date"]
].agg(["min", "max"])

display(date_summary)

,Order Date,Ship Date
min,2014-01-02,2014-01-15
max,2017-12-30,2018-05-01


Ship Date deserves attention!

## Shipping Duration Validation

We calculate the number of days between Order Date and Ship Date to identify potentially abnormal shipping durations.

In [31]:
shipping_days = (
    sales_df["Ship Date"] - sales_df["Order Date"]
).dt.days

shipping_summary = shipping_days.agg(
    ["min", "max", "mean", "median"]
)

display(shipping_summary)

min         0.000000
max       214.000000
mean       34.609065
median      4.000000
dtype: float64

In [32]:
long_shipments = sales_df.loc[
    shipping_days > 30,
    [
        "Order ID",
        "Order Date",
        "Ship Date",
        "Ship Mode",
        "Sales",
        "Profit"
    ]
].copy()

long_shipments["Shipping Days"] = shipping_days[
    shipping_days > 30
]

print(f"Shipments taking more than 30 days: {len(long_shipments):,}")

display(
    long_shipments.sort_values(
        "Shipping Days",
        ascending=False
    ).head(10)
)

Shipments taking more than 30 days: 2,924


,Order ID,Order Date,Ship Date,Ship Mode,Sales,Profit,Shipping Days
920,CA-2014-138296,2014-05-12,2014-12-12,Standard Class,24.560,6.8768,214
2436,CA-2017-140053,2017-03-07,2017-10-07,Standard Class,545.850,114.6285,214
8893,CA-2014-124723,2014-05-08,2014-12-08,Standard Class,489.230,41.9340,214
2771,CA-2016-163986,2016-03-09,2016-10-09,Standard Class,54.500,14.1700,214
2585,CA-2015-121041,2015-03-11,2015-10-11,Standard Class,6.608,2.1476,214
3861,CA-2014-135755,2014-03-05,2014-10-05,Standard Class,40.176,14.5638,214
3862,CA-2014-135755,2014-03-05,2014-10-05,Standard Class,10.896,3.9498,214
3473,CA-2017-111815,2017-03-03,2017-10-03,Standard Class,180.980,47.0548,214
3474,CA-2017-111815,2017-03-03,2017-10-03,Standard Class,99.980,42.9914,214
837,US-2017-156083,2017-04-11,2017-11-11,Standard Class,9.664,3.2616,214


In [33]:
shipping_distribution = (
    shipping_days
    .value_counts()
    .sort_index()
)

display(shipping_distribution.head(20))

0     2227
1      205
2      731
3      531
4     1380
5     1080
6      554
7      273
28      12
29       3
30      74
31      75
32      12
33      25
34      75
35      33
36      44
37      13
59      82
60      32
Name: count, dtype: int64

In [34]:
print("Shipping days frequency:")
display(
    shipping_distribution
    .sort_values(ascending=False)
    .head(30)
)

Shipping days frequency:


0      2227
4      1380
5      1080
2       731
6       554
3       531
153     415
122     396
61      276
7       273
92      210
1       205
123     195
120     173
59       82
31       75
91       75
34       75
30       74
183      72
62       71
184      67
121      60
181      60
151      52
150      50
89       44
36       44
214      38
212      37
Name: count, dtype: int64

In [35]:
shipping_by_mode = (
    sales_df.assign(
        Shipping_Days=shipping_days
    )
    .groupby("Ship Mode")["Shipping_Days"]
    .agg(
        ["count", "min", "median", "mean", "max"]
    )
)

display(shipping_by_mode)

,count,min,median,mean,max
Ship Mode,,,,,
First Class,1538,0,3.0,23.537711,92
Same Day,543,0,0.0,0.880295,31
Second Class,1945,0,4.0,30.559383,153
Standard Class,5968,0,5.0,41.850871,214


## Establish the Normal Shipping Pattern

Let's calculate the distribution for the short-duration transactions first.

In [36]:
shipping_mode_normal = (
    sales_df.assign(
        Shipping_Days=shipping_days
    )
    .loc[lambda df: df["Shipping_Days"] <= 7]
    .groupby("Ship Mode")["Shipping_Days"]
    .agg(
        ["count", "min", "median", "mean", "max"]
    )
)

display(shipping_mode_normal)

,count,min,median,mean,max
Ship Mode,,,,,
First Class,987,0,2.0,1.820669,4
Same Day,527,0,0.0,0.011385,1
Second Class,1341,0,2.0,2.431022,5
Standard Class,4126,0,4.0,3.478429,7


In [37]:
shipping_mode_compliance = (
    sales_df.assign(
        Shipping_Days=shipping_days
    )
    .groupby("Ship Mode")
    .agg(
        Total_Records=("Shipping_Days", "size"),
        Within_7_Days=("Shipping_Days", lambda x: (x <= 7).sum())
    )
)

shipping_mode_compliance["Within_7_Days_%"] = (
    shipping_mode_compliance["Within_7_Days"]
    / shipping_mode_compliance["Total_Records"]
    * 100
)

display(shipping_mode_compliance)

,Total_Records,Within_7_Days,Within_7_Days_%
Ship Mode,,,
First Class,1538,987,64.174252
Same Day,543,527,97.053407
Second Class,1945,1341,68.946015
Standard Class,5968,4126,69.135389


**Finding:** Shipping durations show a strong concentration within 0–7 days across all shipping modes, with medians ranging from 0 to 4 days. However, a substantial number of records contain much longer durations, including repeated clusters around approximately 30, 60, 90, 120, 150, 180, and 210+ days. The maximum duration is 214 days. These patterns are inconsistent with the dominant operational shipping pattern and indicate a potential systematic issue in the source Ship Date values. The dates will be investigated further before being used for shipping-performance analysis.

## Anomaly in Shipping Dates

Let's test whether anomalous Ship Dates preserve the same day of month as Order Dates.

In [38]:
anomalous_shipping = sales_df.assign(
    Shipping_Days=shipping_days
)

anomalous_shipping = anomalous_shipping[
    anomalous_shipping["Shipping_Days"] > 7
].copy()

anomalous_shipping["Same_Day_of_Month"] = (
    anomalous_shipping["Order Date"].dt.day
    ==
    anomalous_shipping["Ship Date"].dt.day
)

print(
    f"Anomalous shipments: {len(anomalous_shipping):,}"
)

print(
    f"Same day-of-month: "
    f"{anomalous_shipping['Same_Day_of_Month'].sum():,}"
)

print(
    f"Same day-of-month %: "
    f"{anomalous_shipping['Same_Day_of_Month'].mean() * 100:.2f}%"
)

Anomalous shipments: 3,013
Same day-of-month: 2,510
Same day-of-month %: 83.31%


**Final Assessment:** 3,013 transactions have shipping durations greater than 7 days, and 83.31% of these anomalous records preserve the same day of month between Order Date and Ship Date. Combined with the repeated clusters of approximately 30, 60, 90, 120, 150, 180, and 210+ days, this indicates a systematic anomaly in the source Ship Date values rather than a reliable representation of actual delivery duration.

The raw Ship Date will therefore be preserved without arbitrary correction. A derived shipping-duration field may be created later with an explicit anomaly flag, but delivery-performance KPIs will not be based on unvalidated corrected dates.

# 2. Transformation Strategy

The raw datasets will be transformed into a star-schema analytical model.

# 2.1 Target Model

- **FactSales:** Transaction-line-level sales records and numeric business measures.
- **DimDate:** One record per calendar date for time-based analysis.
- **DimCustomer:** Stable customer attributes including Customer ID, Customer Name, and Segment.
- **DimProduct:** Product identifier and descriptive product attributes, with known source-data mapping inconsistencies documented.
- **DimGeography:** Transaction-level geographic attributes including City, State, Postal Code, and Region.

### Transformation Responsibilities

**Python**
- Source-data profiling
- Data-quality investigation
- Structural validation

**Power Query**
- Data type enforcement
- Column selection and renaming
- Derived fields
- Dimension construction
- Data preparation for the analytical model

**DAX**
- Business measures
- Time intelligence
- Ratios and profitability metrics
- Growth and ranking calculations

The objective is to keep transformations in the layer where they are most appropriate rather than performing all preparation in a single tool.

# 2.2 Column Classification

Each raw field is classified according to its role in the target analytical model.

The objective is to avoid carrying unnecessary columns into the semantic model while preserving fields required for analysis, relationships, filtering, and business calculations.

## Salesperson Attribute Assessment

We assess the cardinality and transaction coverage of the Retail Sales People field before deciding whether it warrants a dedicated dimension in the analytical model.

In [39]:
salesperson_summary = (
    sales_df["Retail Sales People"]
    .value_counts()
    .rename_axis("Retail Sales People")
    .reset_index(name="Transaction Lines")
)

salesperson_summary["Transaction Share %"] = (
    salesperson_summary["Transaction Lines"]
    / len(sales_df)
    * 100
)

display(salesperson_summary)

,Retail Sales People,Transaction Lines,Transaction Share %
0,Anna Andreadi,3203,32.049230
1,Chuck Magee,2848,28.497098
2,Kelly Williams,2323,23.243946
3,Cassandra Brandow,1620,16.209726


**Finding:** The `Retail Sales People` field contains four distinct salespeople, with all 9,994 transaction lines assigned to a salesperson. The field therefore represents a meaningful business dimension and will be modeled as a dedicated `DimSalesperson` table.

## Final Column Classification

| Raw Column | Target Model | Role |
|---|---|---|
| `Row ID` | FactSales | Transaction-line identifier |
| `Order ID` | FactSales | Order identifier |
| `Order Date` | FactSales | Transaction date |
| `Ship Date` | FactSales | Shipping date |
| `Ship Mode` | FactSales | Shipping method |
| `Customer ID` | FactSales | Customer foreign key |
| `Customer Name` | DimCustomer | Customer attribute |
| `Segment` | DimCustomer | Customer attribute |
| `Country` | DimGeography | Geographic attribute |
| `City` | DimGeography | Geographic attribute |
| `State` | DimGeography | Geographic attribute |
| `Postal Code` | DimGeography | Geographic identifier |
| `Region` | DimGeography | Geographic attribute |
| `Retail Sales People` | DimSalesperson | Salesperson attribute |
| `Product ID` | FactSales | Product foreign key |
| `Category` | DimProduct | Product hierarchy |
| `Sub-Category` | DimProduct | Product hierarchy |
| `Product Name` | DimProduct | Product description |
| `Returned` | FactSales | Transaction status |
| `Sales` | FactSales | Sales measure |
| `Quantity` | FactSales | Quantity measure |
| `Discount` | FactSales | Discount measure |
| `Profit` | FactSales | Profit measure |

## Calendar Table Validation

The calendar table will serve as the Date dimension for the Power BI model. Before incorporating it into the analytical model, we validate that it provides continuous daily coverage across the transactional period.

In [40]:
# Date Continuity Check
calendar_dates = calendar_df["Date"].sort_values()

date_gaps = (
    calendar_dates
    .diff()
    .dt.days
    .value_counts()
    .sort_index()
)

display(date_gaps)

Date
1.0    1825
Name: count, dtype: int64

In [41]:
# Calendar Coverage vs Sales

print("Sales Order Date range:")
print(
    sales_df["Order Date"].min(),
    "→",
    sales_df["Order Date"].max()
)

print("\nCalendar Date range:")
print(
    calendar_df["Date"].min(),
    "→",
    calendar_df["Date"].max()
)

Sales Order Date range:
2014-01-02 00:00:00 → 2017-12-30 00:00:00

Calendar Date range:
2014-01-01 00:00:00 → 2018-12-31 00:00:00


In [42]:
order_dates = set(sales_df["Order Date"].dt.normalize())
calendar_dates_set = set(calendar_df["Date"].dt.normalize())

missing_from_calendar = order_dates - calendar_dates_set

print(
    f"Order dates missing from calendar: "
    f"{len(missing_from_calendar):,}"
)

Order dates missing from calendar: 0


**Finding:** The calendar table provides continuous daily coverage from January 1, 2014 through December 31, 2018. All transaction Order Dates are covered by the calendar, with zero missing dates. The calendar is therefore suitable for use as the model's Date dimension.

Our final `DimDate`:

- Date
- Year
- Quarter
- Quarter (Q)
- Quarter & Year
- Month
- Month Name
- Month & Year
- Week of Year
- Week of Year (W)
- Day of Week
- Day Name

# 3. Target Analytical Model

The final Power BI model will use a star-schema architecture consisting of one transaction fact table and supporting dimensions.

We now have a defined target model:

```text
                    DimDate
                       │
                       │
                       ▼
DimCustomer ─────── FactSales ─────── DimProduct
                       │
              ┌────────┴────────┐
              ▼                 ▼
       DimGeography       DimSalesperson
```
### Fact Table

**FactSales**
- Row ID
- Order ID
- Order Date
- Ship Date
- Ship Mode
- Customer ID
- Product ID
- Geography Key
- Salesperson Key
- Returned
- Sales
- Quantity
- Discount
- Profit

### Dimension Tables

**DimDate**
- Date
- Year
- Quarter
- Quarter (Q)
- Quarter & Year
- Month
- Month Name
- Month & Year
- Week of Year
- Week of Year (W)
- Day of Week
- Day Name

**DimCustomer**
- Customer ID
- Customer Name
- Segment

**DimProduct**
- Product ID
- Product Name
- Category
- Sub-Category

**DimGeography**
- Geography Key
- Country
- City
- State
- Postal Code
- Region

**DimSalesperson**
- Salesperson Key
- Retail Sales People

### Modeling Principles

- FactSales remains at transaction-line grain.
- Dimensions contain descriptive attributes used for filtering and grouping.
- Business measures will be implemented in DAX rather than hard-coded into the source data.
- Raw source fields will not be arbitrarily overwritten.
- Known source-data anomalies will be documented and handled explicitly.

Note: We haven't actually created `Geography Key` or `Salesperson` Key yet. Those are target-model fields, not current source columns. We'll create them in Power Query when we build the dimensions.

---


<p style="text-align:center; color:skyblue; font-size:18px;">
© 2026 Mostafizur Rahman
</p>
